### PR TITLE
Fix next phase for next build

### LINKED ISSUE #57927: [Next14] Phase in next config is set to `PHASE_DEVELOPMENT_SERVER` when running `next build`
### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/great-cdn-k9f434

### To Reproduce

1. Run the build script in the linked reproduction sandbox
2. Observe the logged output for the phase values

### Current vs. Expected behavior

**Current behavior**
Starting with next.js v14, it seems that the next config is loaded twice when running `next build`:

- The first time, the `phase` value is set to `PHASE_PRODUCTION_BUILD` and the invocation is triggered by https://github.com/vercel/next.js/blob/1c5d7782a02259007a101124f6774d0251c7d3ff/packages/next/src/build/index.ts#L372
- The second time, the `phase` value is set to `PHASE_DEVELOPMENT_SERVER` and the invocation is triggered by `getStartServerInfo` in https://github.com/vercel/next.js/blob/1c5d7782a02259007a101124f6774d0251c7d3ff/packages/next/src/build/index.ts#L514. It seems that `getStartServerInfo` always uses `PHASE_DEVELOPMENT_SERVER`, no matter the context: https://github.com/vercel/next.js/blob/1c5d7782a02259007a101124f6774d0251c7d3ff/packages/next/src/server/lib/app-info-log.ts#L58

**Expected behavior**
`getStartServerInfo` should be called with a value of `PHASE_PRODUCTION_BUILD` when run in a `next build` context.

In my case, I used the `phase` value to skip the execution of some logic only in the buildphase which is now no longer possible.
I worked around this issue by using a custom env var when launching the build`BUILD_PHASE=true next build` and then checking for the value of this env var in my `next.config.js`. 

I wonder if this change is intended behaviour or a bug. In my opinion, the `phase` argument loses most of its usefulness if it's set this way.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 22.6.0: Fri Sep 15 13:39:52 PDT 2023; root:xnu-8796.141.3.700.8~1/RELEASE_X86_64
Binaries:
  Node: 20.9.0
  npm: 10.1.0
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 14.0.2-canary.6
  eslint-config-next: 14.0.1
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Additional context

I use a custom server for my next.js app. However, the example repo I linked for reproduction suffers from the same issue and does not use a custom server.

<sub>[NEXT-2179](https://linear.app/vercel/issue/NEXT-2179/[next14]-phase-in-next-config-is-set-to-phase-development-server-when)</sub>

### PR DESCRIPTION
**Pull request by huozhi**
In both dev server and production build we both use `getStartServerInfo` to log the basic info but for prod build we should always respect to use "build" phase 

Fixes #57927 
Closes NEXT-2179

### CODE REVIEW

File: test/e2e/next-phase/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      expect(next.cliOutput).toContain(currentPhase)
+      expect(next.cliOutput).not.toContain(nonExistedPhase)
+
+      await next.fetch('/')
+      await next.fetch('/foo')
+
+      if (isNextDev) {
+        expect(next.cliOutput).not.toContain('phase-production-server')
+      } else {
+        expect(next.cliOutput).toContain('phase-production-server')
```
Comment (timneutkens) [inline test/e2e/next-phase/index.test.ts:38]:
```suggestion
        expect(next.cliOutput).toContain(phases.build)
```

File: test/e2e/next-phase/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      const currentPhase = isNextDev ? phases.dev : phases.build
+      const nonExistedPhase = isNextDev ? phases.build : phases.dev
+
+      expect(next.cliOutput).toContain(currentPhase)
+      expect(next.cliOutput).not.toContain(nonExistedPhase)
+
+      await next.fetch('/')
+      await next.fetch('/foo')
+
+      if (isNextDev) {
```
Comment (timneutkens) [inline test/e2e/next-phase/index.test.ts:35]: This condition looks a bit off as you already checked `toContain` earlier. Can you add a comment here to explain that it's checking if it's not being logged.

File: test/e2e/next-phase/index.test.ts (hunk 1 above)
Comment (timneutkens) [reply test/e2e/next-phase/index.test.ts:38]: The condition looks unexpected btw, as it was already true before, you probably want to check for dev again?

```suggestion
        expect(next.cliOutput).not.toContain(phases.dev)
```

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| buildDuration | 12.9s | 13.1s | ⚠️ +190ms |
| buildDurationCached | 6.7s | 5.8s | N/A |
| nodeModulesSize | 200 MB | 200 MB | ⚠️ +614 B |
| nextStartRea..uration (ms) | 440ms | 440ms | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| 193.HASH.js gzip | 181 B | 182 B | N/A |
| 3f784ff6-HASH.js gzip | 53.4 kB | 53.4 kB | ✓ |
| 433-HASH.js gzip | 29 kB | 29 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 239 B | 242 B | N/A |
| main-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 98.6 kB | 98.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 183 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 502 B | N/A |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 253 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.18 kB | 4.18 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | ✓ |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.4 kB | 3.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| index.html gzip | 529 B | 528 B | N/A |
| link.html gzip | 541 B | 540 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94 kB | 94 kB | N/A |
| page.js gzip | 149 kB | 149 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 625 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.5 kB | 37.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/build-phase | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 169 kB | 169 kB | ✓ |
| app-page-exp..prod.js gzip | 95.6 kB | 95.6 kB | ✓ |
| app-page-tur..prod.js gzip | 96.3 kB | 96.3 kB | ✓ |
| app-page-tur..prod.js gzip | 90.9 kB | 90.9 kB | ✓ |
| app-page.run...dev.js gzip | 142 kB | 142 kB | ✓ |
| app-page.run..prod.js gzip | 90.2 kB | 90.2 kB | ✓ |
| app-route-ex...dev.js gzip | 24.2 kB | 24.2 kB | ✓ |
| app-route-ex..prod.js gzip | 16.9 kB | 16.9 kB | ✓ |
| app-route-tu..prod.js gzip | 16.9 kB | 16.9 kB | ✓ |
| app-route-tu..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route.ru...dev.js gzip | 23.6 kB | 23.6 kB | ✓ |
| app-route.ru..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.41 kB | 9.41 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.68 kB | 9.68 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.4 kB | 9.4 kB | ✓ |
| pages-turbo...prod.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim...dev.js gzip | 22.6 kB | 22.6 kB | ✓ |
| pages.runtim..prod.js gzip | 22 kB | 22 kB | ✓ |
| server.runti..prod.js gzip | 49.7 kB | 49.7 kB | ✓ |
| Overall change | 944 kB | 944 kB | ✓ |

</details>

</details>
Commit: 56e08df03b10a4ff9d0dada39563b2042c2258f2

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 215417bccaa03397a79bd42352ef735df90bc936

`pnpm test-dev test/development/app-hmr/hmr.test.ts` (PPR)
- app-dir-hmr > filesystem changes > should not break when renaming a folder

<details>
<summary>Expand output</summary>

  ● app-dir-hmr › filesystem changes › should not break when renaming a folder

    TIMED OUT: success

    undefined

    Error: elementHandle.innerText: Element is not attached to the DOM

      636 |
      637 |   if (hardError) {
    > 638 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      639 |   }
      640 |   return false
      641 | }

      at check (lib/next-test-utils.ts:638:11)
      at Object.<anonymous> (development/app-hmr/hmr.test.ts:44:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by timneutkens — APPROVED** [review]

**Pull request merged**